[faiss] Unable to create knn index from model

When a user wants to create a k-NN index from a model, they need to pass a mapping like:
```
PUT /target_index
{
  "settings" : {
    "number_of_shards" : 3,
    "number_of_replicas" : 1,
    "index.knn": true
  },
  "mappings": {
       "properties": {
       "target_field": {
           "type": "knn_vector",
           "model_id": "my_model"
      }
   }
  }
}
```

Internally, we require other information about the model when initializing the index, such as dimension, engine and space type. Not to mention, we need to check if the model exists. 

From the [KNNVectorFieldMapper](https://github.com/opensearch-project/k-NN/blob/feature/faiss-support/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java#L90), we cannot get that information from the model system index directly. To get around this, we added metadata to the model system index. Then, we [retrieve this during mapper creation](https://github.com/opensearch-project/k-NN/blob/feature/faiss-support/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java#L236). 

The issue with this approach is that sometimes the [cluster state](https://github.com/opensearch-project/k-NN/blob/feature/faiss-support/src/main/java/org/opensearch/knn/indices/ModelDao.java#L334) is not available when this build function is called. For instance, sometimes the mapper is created in the same thread where the state is being updated. The resulting error looks something like this:
```
=== Standard error of node `node{::integTest-0}` ===
?   ? last 40 non error or warning messages from /home/test/k-NN-1/build/testclusters/integTest-0/logs/opensearch.stderr.log ?
? WARNING clustering 1001 points to 128 centroids: please provide at least 4992 training points
?  fatal error in thread [opensearch[integTest-0][clusterApplierService#updateTask][T#1]], exiting
?  java.lang.AssertionError: should not be called by a cluster state applier. reason [the applied cluster state is not yet available]
```

From my understanding, there is no safe way to access the cluster state from the field mapper builder. For us, that means we cannot get the dimension, space type, and engine (as well as check if the model exists) during index creation. Instead, the next best thing we can do is validate this information during indexing, from the [parseCreateField](https://github.com/opensearch-project/k-NN/blob/feature/faiss-support/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java#L383) method.

--- a/src/main/java/org/opensearch/knn/index/KNNQueryBuilder.java
+++ b/src/main/java/org/opensearch/knn/index/KNNQueryBuilder.java
@@ -25,6 +25,8 @@
 
 package org.opensearch.knn.index;
 
+import org.opensearch.knn.indices.ModelDao;
+import org.opensearch.knn.indices.ModelMetadata;
 import org.opensearch.knn.plugin.stats.KNNCounter;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -211,6 +213,22 @@ public class KNNQueryBuilder extends AbstractQueryBuilder<KNNQueryBuilder> {
         }
 
         int dimension = ((KNNVectorFieldMapper.KNNVectorFieldType) mappedFieldType).getDimension();
+
+        //TODO: Fix this logic
+        if (dimension == -1) {
+            String modelId = ((KNNVectorFieldMapper.KNNVectorFieldType) mappedFieldType).getModelId();
+
+            if (modelId == null) {
+                throw new IllegalArgumentException("Field does not have dimension set.");
+            }
+
+            ModelMetadata modelMetadata = ModelDao.OpenSearchKNNModelDao.getInstance().getMetadata(modelId);
+
+            if (modelMetadata == null) {
+                throw new IllegalArgumentException("Model ID \"" + modelId + "\" does not exist.");
+            }
+            dimension = modelMetadata.getDimension();
+        }
 
         if (dimension != vector.length) {
             throw new IllegalArgumentException("Query vector has invalid dimension: " + vector.length +

--- a/src/main/java/org/opensearch/knn/index/KNNQueryBuilder.java
+++ b/src/main/java/org/opensearch/knn/index/KNNQueryBuilder.java
@@ -51,6 +51,8 @@ import java.util.Objects;
  */
 public class KNNQueryBuilder extends AbstractQueryBuilder<KNNQueryBuilder> {
     private static Logger logger = LogManager.getLogger(KNNQueryBuilder.class);
+    private static ModelDao modelDao;
+
     public static final ParseField VECTOR_FIELD = new ParseField("vector");
     public static final ParseField K_FIELD = new ParseField("k");
     public static int K_MAX = 10000;
@@ -92,6 +94,10 @@ public class KNNQueryBuilder extends AbstractQueryBuilder<KNNQueryBuilder> {
         this.fieldName = fieldName;
         this.vector = vector;
         this.k = k;
+    }
+
+    public static void initialize(ModelDao modelDao) {
+        KNNQueryBuilder.modelDao = modelDao;
     }
 
     private static float[] ObjectsToFloats(List<Object> objs) {
@@ -214,15 +220,15 @@ public class KNNQueryBuilder extends AbstractQueryBuilder<KNNQueryBuilder> {
 
         int dimension = ((KNNVectorFieldMapper.KNNVectorFieldType) mappedFieldType).getDimension();
 
-        //TODO: Fix this logic
+        // If the dimension is not set, then the only valid route forward is if the field uses a model
         if (dimension == -1) {
             String modelId = ((KNNVectorFieldMapper.KNNVectorFieldType) mappedFieldType).getModelId();
 
             if (modelId == null) {
-                throw new IllegalArgumentException("Field does not have dimension set.");
+                throw new IllegalArgumentException("Field '" + this.fieldName + "' does not have dimension set.");
             }
 
-            ModelMetadata modelMetadata = ModelDao.OpenSearchKNNModelDao.getInstance().getMetadata(modelId);
+            ModelMetadata modelMetadata = modelDao.getMetadata(modelId);
 
             if (modelMetadata == null) {
                 throw new IllegalArgumentException("Model ID \"" + modelId + "\" does not exist.");

--- a/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java
+++ b/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java
@@ -233,7 +233,10 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
 
             String modelIdAsString = this.modelId.get();
             if (modelIdAsString != null) {
-                ModelMetadata modelMetadata = modelDao.getMetadata(modelIdAsString);
+                // Because model information is stored in cluster metadata, we are unable to get it here. This is
+                // because to get the cluster metadata, you need access to the cluster state. Because this code is
+                // sometimes used to initialize the cluster state/update cluster state, we cannot get the state here
+                // safely. So, we are unable to validate the model :(. The model gets validated during ingestion.
 
                 if (modelMetadata == null) {
                     throw new IllegalArgumentException("Model \"" + modelId + "\" does not exist");
@@ -241,15 +244,15 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
 
                 return new ModelFieldMapper(
                         name,
-                        new KNNVectorFieldType(buildFullName(context), meta.getValue(), modelMetadata.getDimension()),
+                        //TODO: We shall see if this works
+                        new KNNVectorFieldType(buildFullName(context), meta.getValue(), -1, modelIdAsString),
                         multiFieldsBuilder.build(this, context),
                         copyTo.build(),
                         ignoreMalformed(context),
                         stored.get(),
                         hasDocValues.get(),
                         modelDao,
-                        modelIdAsString,
-                        modelMetadata);
+                        modelIdAsString);
             }
 
             // Build legacy
@@ -314,10 +317,18 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
     public static class KNNVectorFieldType extends MappedFieldType {
 
         int dimension;
+        String modelId;
 
         public KNNVectorFieldType(String name, Map<String, String> meta, int dimension) {
             super(name, false, false, true, TextSearchInfo.NONE, meta);
             this.dimension = dimension;
+            this.modelId = null; // By default, this is null. It will only be set for a mapping with a modelId set.
+        }
+
+        public KNNVectorFieldType(String name, Map<String, String> meta, int dimension, String modelId) {
+            super(name, false, false, true, TextSearchInfo.NONE, meta);
+            this.dimension = dimension;
+            this.modelId = modelId;
         }
 
         @Override
@@ -343,6 +354,10 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
 
         public int getDimension() {
             return dimension;
+        }
+
+        public String getModelId() {
+            return modelId;
         }
 
         @Override
@@ -604,23 +619,90 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
      */
     protected static class ModelFieldMapper extends KNNVectorFieldMapper {
 
+        //TODO: We need to make sure dimension doesnt show up in mapping as -1.
         private ModelFieldMapper(String simpleName, KNNVectorFieldType mappedFieldType, MultiFields multiFields,
                                 CopyTo copyTo, Explicit<Boolean> ignoreMalformed, boolean stored,
-                                boolean hasDocValues, ModelDao modelDao, String modelId, ModelMetadata modelMetadata) {
+                                boolean hasDocValues, ModelDao modelDao, String modelId) {
             super(simpleName, mappedFieldType, multiFields, copyTo, ignoreMalformed, stored, hasDocValues);
 
             this.modelId = modelId;
             this.modelDao = modelDao;
 
             this.fieldType = new FieldType(KNNVectorFieldMapper.Defaults.FIELD_TYPE);
-
             this.fieldType.putAttribute(MODEL_ID, modelId);
-
-            this.fieldType.putAttribute(DIMENSION, String.valueOf(modelMetadata.getDimension()));
-            this.fieldType.putAttribute(SPACE_TYPE, modelMetadata.getSpaceType().getValue());
-            this.fieldType.putAttribute(KNN_ENGINE, modelMetadata.getKnnEngine().getName());
-
             this.fieldType.freeze();
+        }
+
+        @Override
+        protected void parseCreateField(ParseContext context) throws IOException {
+            if (!KNNSettings.isKNNPluginEnabled()) {
+                throw new IllegalStateException("KNN plugin is disabled. To enable " +
+                        "update knn.plugin.enabled setting to true");
+            }
+
+            if (KNNSettings.isCircuitBreakerTriggered()) {
+                throw new IllegalStateException("Indexing knn vector fields is rejected as circuit breaker triggered." +
+                        " Check _opendistro/_knn/stats for detailed state");
+            }
+
+            //TODO: Because we cannot do validation above, we need to do it here
+            ModelMetadata modelMetadata = this.modelDao.getMetadata(modelId);
+
+            context.path().add(simpleName());
+
+            ArrayList<Float> vector = new ArrayList<>();
+            XContentParser.Token token = context.parser().currentToken();
+            float value;
+            if (token == XContentParser.Token.START_ARRAY) {
+                token = context.parser().nextToken();
+                while (token != XContentParser.Token.END_ARRAY) {
+                    value = context.parser().floatValue();
+
+                    if (Float.isNaN(value)) {
+                        throw new IllegalArgumentException("KNN vector values cannot be NaN");
+                    }
+
+                    if (Float.isInfinite(value)) {
+                        throw new IllegalArgumentException("KNN vector values cannot be infinity");
+                    }
+
+                    vector.add(value);
+                    token = context.parser().nextToken();
+                }
+            } else if (token == XContentParser.Token.VALUE_NUMBER) {
+                value = context.parser().floatValue();
+
+                if (Float.isNaN(value)) {
+                    throw new IllegalArgumentException("KNN vector values cannot be NaN");
+                }
+
+                if (Float.isInfinite(value)) {
+                    throw new IllegalArgumentException("KNN vector values cannot be infinity");
+                }
+
+                vector.add(value);
+                context.parser().nextToken();
+            }
+
+            if (modelMetadata.getDimension() != vector.size()) {
+                String errorMessage = String.format("Vector dimension mismatch. Expected: %d, Given: %d",
+                        modelMetadata.getDimension(), vector.size());
+                throw new IllegalArgumentException(errorMessage);
+            }
+
+            float[] array = new float[vector.size()];
+            int i = 0;
+            for (Float f : vector) {
+                array[i++] = f;
+            }
+
+            VectorField point = new VectorField(name(), array, fieldType);
+
+            context.doc().add(point);
+            if (fieldType.stored()) {
+                context.doc().add(new StoredField(name(), point.toString()));
+            }
+            context.path().remove();
         }
     }
 }

--- a/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java
+++ b/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java
@@ -638,8 +638,10 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
             ModelMetadata modelMetadata = this.modelDao.getMetadata(modelId);
 
             if (modelMetadata == null) {
-                throw new IllegalStateException("Model \"" + modelId + "\" does not exist. A new index will need to " +
-                        "be created.");
+                throw new IllegalStateException("Model \"" + modelId + "\" from " +
+                        context.mapperService().index().getName() + "'s mapping does not exist. Because the " +
+                        "\"" + MODEL_ID + "\" parameter is not updateable, this index will need to " +
+                        "be recreated with a valid model.");
             }
 
             parseCreateField(context, modelMetadata.getDimension());

--- a/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java
+++ b/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java
@@ -236,7 +236,7 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
                 // Because model information is stored in cluster metadata, we are unable to get it here. This is
                 // because to get the cluster metadata, you need access to the cluster state. Because this code is
                 // sometimes used to initialize the cluster state/update cluster state, we cannot get the state here
-                // safely. So, we are unable to validate the model :(. The model gets validated during ingestion.
+                // safely. So, we are unable to validate the model. The model gets validated during ingestion.
 
                 if (modelMetadata == null) {
                     throw new IllegalArgumentException("Model \"" + modelId + "\" does not exist");
@@ -244,7 +244,6 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
 
                 return new ModelFieldMapper(
                         name,
-                        //TODO: We shall see if this works
                         new KNNVectorFieldType(buildFullName(context), meta.getValue(), -1, modelIdAsString),
                         multiFieldsBuilder.build(this, context),
                         copyTo.build(),
@@ -400,6 +399,11 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
 
     @Override
     protected void parseCreateField(ParseContext context) throws IOException {
+        parseCreateField(context, fieldType().getDimension());
+    }
+
+    protected void parseCreateField(ParseContext context, int dimension) throws IOException {
+
         if (!KNNSettings.isKNNPluginEnabled()) {
             throw new IllegalStateException("KNN plugin is disabled. To enable " +
                     "update knn.plugin.enabled setting to true");
@@ -446,9 +450,9 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
             context.parser().nextToken();
         }
 
-        if (fieldType().dimension != vector.size()) {
-            String errorMessage = String.format("Vector dimension mismatch. Expected: %d, Given: %d",
-                    fieldType().dimension, vector.size());
+        if (dimension != vector.size()) {
+            String errorMessage = String.format("Vector dimension mismatch. Expected: %d, Given: %d", dimension,
+                    vector.size());
             throw new IllegalArgumentException(errorMessage);
         }
 
@@ -619,7 +623,6 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
      */
     protected static class ModelFieldMapper extends KNNVectorFieldMapper {
 
-        //TODO: We need to make sure dimension doesnt show up in mapping as -1.
         private ModelFieldMapper(String simpleName, KNNVectorFieldType mappedFieldType, MultiFields multiFields,
                                 CopyTo copyTo, Explicit<Boolean> ignoreMalformed, boolean stored,
                                 boolean hasDocValues, ModelDao modelDao, String modelId) {
@@ -635,74 +638,17 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
 
         @Override
         protected void parseCreateField(ParseContext context) throws IOException {
-            if (!KNNSettings.isKNNPluginEnabled()) {
-                throw new IllegalStateException("KNN plugin is disabled. To enable " +
-                        "update knn.plugin.enabled setting to true");
-            }
-
-            if (KNNSettings.isCircuitBreakerTriggered()) {
-                throw new IllegalStateException("Indexing knn vector fields is rejected as circuit breaker triggered." +
-                        " Check _opendistro/_knn/stats for detailed state");
-            }
-
-            //TODO: Because we cannot do validation above, we need to do it here
+            // For the model field mapper, we cannot validate the model during index creation due to
+            // an issue with reading cluster state during mapper creation. So, we need to validate the
+            // model when ingestion starts.
             ModelMetadata modelMetadata = this.modelDao.getMetadata(modelId);
 
-            context.path().add(simpleName());
-
-            ArrayList<Float> vector = new ArrayList<>();
-            XContentParser.Token token = context.parser().currentToken();
-            float value;
-            if (token == XContentParser.Token.START_ARRAY) {
-                token = context.parser().nextToken();
-                while (token != XContentParser.Token.END_ARRAY) {
-                    value = context.parser().floatValue();
-
-                    if (Float.isNaN(value)) {
-                        throw new IllegalArgumentException("KNN vector values cannot be NaN");
-                    }
-
-                    if (Float.isInfinite(value)) {
-                        throw new IllegalArgumentException("KNN vector values cannot be infinity");
-                    }
-
-                    vector.add(value);
-                    token = context.parser().nextToken();
-                }
-            } else if (token == XContentParser.Token.VALUE_NUMBER) {
-                value = context.parser().floatValue();
-
-                if (Float.isNaN(value)) {
-                    throw new IllegalArgumentException("KNN vector values cannot be NaN");
-                }
-
-                if (Float.isInfinite(value)) {
-                    throw new IllegalArgumentException("KNN vector values cannot be infinity");
-                }
-
-                vector.add(value);
-                context.parser().nextToken();
+            if (modelMetadata == null) {
+                throw new IllegalStateException("Model \"" + modelId + "\" does not exist. A new index will need to " +
+                        "be created.");
             }
 
-            if (modelMetadata.getDimension() != vector.size()) {
-                String errorMessage = String.format("Vector dimension mismatch. Expected: %d, Given: %d",
-                        modelMetadata.getDimension(), vector.size());
-                throw new IllegalArgumentException(errorMessage);
-            }
-
-            float[] array = new float[vector.size()];
-            int i = 0;
-            for (Float f : vector) {
-                array[i++] = f;
-            }
-
-            VectorField point = new VectorField(name(), array, fieldType);
-
-            context.doc().add(point);
-            if (fieldType.stored()) {
-                context.doc().add(new StoredField(name(), point.toString()));
-            }
-            context.path().remove();
+            parseCreateField(context, modelMetadata.getDimension());
         }
     }
 }

--- a/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java
+++ b/src/main/java/org/opensearch/knn/index/KNNVectorFieldMapper.java
@@ -238,10 +238,6 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
                 // sometimes used to initialize the cluster state/update cluster state, we cannot get the state here
                 // safely. So, we are unable to validate the model. The model gets validated during ingestion.
 
-                if (modelMetadata == null) {
-                    throw new IllegalArgumentException("Model \"" + modelId + "\" does not exist");
-                }
-
                 return new ModelFieldMapper(
                         name,
                         new KNNVectorFieldType(buildFullName(context), meta.getValue(), -1, modelIdAsString),
@@ -319,9 +315,7 @@ public abstract class KNNVectorFieldMapper extends ParametrizedFieldMapper {
         String modelId;
 
         public KNNVectorFieldType(String name, Map<String, String> meta, int dimension) {
-            super(name, false, false, true, TextSearchInfo.NONE, meta);
-            this.dimension = dimension;
-            this.modelId = null; // By default, this is null. It will only be set for a mapping with a modelId set.
+            this(name, meta, dimension, null);
         }
 
         public KNNVectorFieldType(String name, Map<String, String> meta, int dimension, String modelId) {

--- a/src/main/java/org/opensearch/knn/index/KNNWeight.java
+++ b/src/main/java/org/opensearch/knn/index/KNNWeight.java
@@ -71,6 +71,8 @@ import static org.opensearch.knn.plugin.stats.KNNCounter.GRAPH_QUERY_ERRORS;
  */
 public class KNNWeight extends Weight {
     private static Logger logger = LogManager.getLogger(KNNWeight.class);
+    private static ModelDao modelDao;
+
     private final KNNQuery knnQuery;
     private final float boost;
 
@@ -81,6 +83,10 @@ public class KNNWeight extends Weight {
         this.knnQuery = query;
         this.boost = boost;
         this.nativeMemoryCacheManager = NativeMemoryCacheManager.getInstance();
+    }
+
+    public static void initialize(ModelDao modelDao) {
+        KNNWeight.modelDao = modelDao;
     }
 
     @Override
@@ -94,7 +100,6 @@ public class KNNWeight extends Weight {
 
     @Override
     public Scorer scorer(LeafReaderContext context) throws IOException {
-        //TODO: WE NEED TO FIX THIS I GUESS
             SegmentReader reader = (SegmentReader) FilterLeafReader.unwrap(context.reader());
             String directory = ((FSDirectory) FilterDirectory.unwrap(reader.directory())).getDirectory().toString();
 
@@ -109,12 +114,13 @@ public class KNNWeight extends Weight {
             KNNEngine knnEngine;
             SpaceType spaceType;
 
-            // Here is where things will diverge
+            // Check if a modelId exists. If so, the space type and engine will need to be picked up from the model's
+            // metadata.
             String modelId = fieldInfo.getAttribute(MODEL_ID);
             if (modelId != null) {
-                ModelMetadata modelMetadata = ModelDao.OpenSearchKNNModelDao.getInstance().getMetadata(modelId);
+                ModelMetadata modelMetadata = modelDao.getMetadata(modelId);
                 if (modelMetadata == null) {
-                    throw new RuntimeException("ModelMetadata is null");
+                    throw new RuntimeException("Model \"" + modelId + "\" does not exist.");
                 }
 
                 knnEngine = modelMetadata.getKnnEngine();

--- a/src/main/java/org/opensearch/knn/index/KNNWeight.java
+++ b/src/main/java/org/opensearch/knn/index/KNNWeight.java
@@ -47,6 +47,8 @@ import org.apache.lucene.store.FSDirectory;
 import org.apache.lucene.store.FilterDirectory;
 import org.apache.lucene.util.DocIdSetBuilder;
 import org.opensearch.common.io.PathUtils;
+import org.opensearch.knn.indices.ModelDao;
+import org.opensearch.knn.indices.ModelMetadata;
 import org.opensearch.knn.plugin.stats.KNNCounter;
 
 import java.io.IOException;
@@ -60,6 +62,7 @@ import java.util.concurrent.ExecutionException;
 import java.util.stream.Collectors;
 
 import static org.opensearch.knn.common.KNNConstants.KNN_ENGINE;
+import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
 import static org.opensearch.knn.common.KNNConstants.SPACE_TYPE;
 import static org.opensearch.knn.plugin.stats.KNNCounter.GRAPH_QUERY_ERRORS;
 
@@ -91,6 +94,7 @@ public class KNNWeight extends Weight {
 
     @Override
     public Scorer scorer(LeafReaderContext context) throws IOException {
+        //TODO: WE NEED TO FIX THIS I GUESS
             SegmentReader reader = (SegmentReader) FilterLeafReader.unwrap(context.reader());
             String directory = ((FSDirectory) FilterDirectory.unwrap(reader.directory())).getDirectory().toString();
 
@@ -102,8 +106,23 @@ public class KNNWeight extends Weight {
                 return null;
             }
 
-            KNNEngine knnEngine = KNNEngine.getEngine(fieldInfo.getAttribute(KNN_ENGINE));
-            SpaceType spaceType = SpaceType.getSpace(fieldInfo.getAttribute(SPACE_TYPE));
+            KNNEngine knnEngine;
+            SpaceType spaceType;
+
+            // Here is where things will diverge
+            String modelId = fieldInfo.getAttribute(MODEL_ID);
+            if (modelId != null) {
+                ModelMetadata modelMetadata = ModelDao.OpenSearchKNNModelDao.getInstance().getMetadata(modelId);
+                if (modelMetadata == null) {
+                    throw new RuntimeException("ModelMetadata is null");
+                }
+
+                knnEngine = modelMetadata.getKnnEngine();
+                spaceType = modelMetadata.getSpaceType();
+            } else {
+                knnEngine = KNNEngine.getEngine(fieldInfo.getAttribute(KNN_ENGINE));
+                spaceType = SpaceType.getSpace(fieldInfo.getAttribute(SPACE_TYPE));
+            }
 
             /*
              * In case of compound file, extension would be <engine-extension> + c otherwise <engine-extension>

--- a/src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java
+++ b/src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java
@@ -62,7 +62,6 @@ import java.security.PrivilegedAction;
 import java.util.HashMap;
 import java.util.Map;
 
-import static org.opensearch.knn.common.KNNConstants.DIMENSION;
 import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
 import static org.opensearch.knn.index.codec.KNNCodecUtil.buildEngineFileName;
 
@@ -91,9 +90,6 @@ class KNN80DocValuesConsumer extends DocValuesConsumer implements Closeable {
     public void addKNNBinaryField(FieldInfo field, DocValuesProducer valuesProducer) throws IOException {
         KNNCounter.GRAPH_INDEX_REQUESTS.increment();
         if (field.attributes().containsKey(KNNVectorFieldMapper.KNN_FIELD)) {
-            // Get engine to be used for indexing
-            String engineName = field.attributes().getOrDefault(KNNConstants.KNN_ENGINE, KNNEngine.DEFAULT.getName());
-            KNNEngine knnEngine = KNNEngine.getEngine(engineName);
 
             // Get values to be indexed
             BinaryDocValues values = valuesProducer.getBinary(field);
@@ -104,44 +100,42 @@ class KNN80DocValuesConsumer extends DocValuesConsumer implements Closeable {
             }
 
             // Create library index either from model or from scratch
-            String engineFileName = buildEngineFileName(state.segmentInfo.name, knnEngine.getLatestBuildVersion(),
-                    field.name, knnEngine.getExtension());
-            String indexPath = Paths.get(((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(),
-                    engineFileName).toString();
-            String tmpEngineFileName = engineFileName + TEMP_SUFFIX;
-            String tempIndexPath = indexPath + TEMP_SUFFIX;
+            String engineFileName;
+            String indexPath;
+            String tmpEngineFileName;
+
             if (field.attributes().containsKey(MODEL_ID)) {
+
                 String modelId = field.attributes().get(MODEL_ID);
                 Model model = ModelCache.getInstance().get(modelId);
 
+                KNNEngine knnEngine = model.getModelMetadata().getKnnEngine();
+
+                engineFileName = buildEngineFileName(state.segmentInfo.name, knnEngine.getLatestBuildVersion(),
+                        field.name, knnEngine.getExtension());
+                indexPath = Paths.get(((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(),
+                        engineFileName).toString();
+                tmpEngineFileName = engineFileName + TEMP_SUFFIX;
+                String tempIndexPath = indexPath + TEMP_SUFFIX;
+
                 if (model.getModelBlob() == null) {
-                    throw new RuntimeException("There is no model with id \"" + modelId + "\"");
-                }
-
-                if (model.getModelMetadata().getKnnEngine() != knnEngine) {
-                    throw new RuntimeException("Model Engine \"" + model.getModelMetadata().getKnnEngine().getName()
-                            + "\" cannot be different than index engine \"" + knnEngine.getName() + "\"");
-                }
-
-                String spaceName = field.getAttribute(KNNConstants.SPACE_TYPE);
-                if (spaceName == null) {
-                    throw new RuntimeException("Space Type cannot be null");
-                }
-
-                SpaceType spaceType = SpaceType.getSpace(spaceName);
-                if (model.getModelMetadata().getSpaceType() != spaceType) {
-                    throw new RuntimeException("Model Space Type \"" + model.getModelMetadata().getSpaceType().getValue()
-                            + "\" cannot be different than index Space Type \"" + spaceType.getValue() + "\"");
-                }
-
-                int dimension = Integer.parseInt(field.attributes().getOrDefault(DIMENSION, "-1"));
-                if (model.getModelMetadata().getDimension() != dimension) {
-                    throw new RuntimeException("Model dimension \"" + model.getModelMetadata().getDimension()
-                            + "\" cannot be different than index dimension \"" + dimension + "\"");
+                    throw new RuntimeException("There is no trained model with id \"" + modelId + "\"");
                 }
 
                 createKNNIndexFromTemplate(model.getModelBlob(), pair, knnEngine, tempIndexPath);
             } else {
+
+                // Get engine to be used for indexing
+                String engineName = field.attributes().getOrDefault(KNNConstants.KNN_ENGINE, KNNEngine.DEFAULT.getName());
+                KNNEngine knnEngine = KNNEngine.getEngine(engineName);
+
+                engineFileName = buildEngineFileName(state.segmentInfo.name, knnEngine.getLatestBuildVersion(),
+                        field.name, knnEngine.getExtension());
+                indexPath = Paths.get(((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(),
+                        engineFileName).toString();
+                tmpEngineFileName = engineFileName + TEMP_SUFFIX;
+                String tempIndexPath = indexPath + TEMP_SUFFIX;
+
                 createKNNIndexFromScratch(field, pair, knnEngine, tempIndexPath);
             }
 

--- a/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java
+++ b/src/main/java/org/opensearch/knn/plugin/KNNPlugin.java
@@ -30,6 +30,7 @@ import org.opensearch.knn.index.KNNQueryBuilder;
 import org.opensearch.knn.index.KNNSettings;
 import org.opensearch.knn.index.KNNVectorFieldMapper;
 
+import org.opensearch.knn.index.KNNWeight;
 import org.opensearch.knn.index.memory.NativeMemoryLoadStrategy;
 import org.opensearch.knn.indices.ModelCache;
 import org.opensearch.knn.indices.ModelDao;
@@ -170,6 +171,8 @@ public class KNNPlugin extends Plugin implements MapperPlugin, SearchPlugin, Act
         ModelCache.initialize(ModelDao.OpenSearchKNNModelDao.getInstance(), clusterService);
         TrainingJobRunner.initialize(threadPool, ModelDao.OpenSearchKNNModelDao.getInstance());
         KNNCircuitBreaker.getInstance().initialize(threadPool, clusterService, client);
+        KNNQueryBuilder.initialize(ModelDao.OpenSearchKNNModelDao.getInstance());
+        KNNWeight.initialize(ModelDao.OpenSearchKNNModelDao.getInstance());
         knnStats = new KNNStats(KNNStatsConfig.KNN_STATS);
         return ImmutableList.of(knnStats);
     }

--- a/src/test/java/org/opensearch/knn/index/KNNCreateIndexFromModelTests.java
+++ b/src/test/java/org/opensearch/knn/index/KNNCreateIndexFromModelTests.java
@@ -1,0 +1,79 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index;
+
+import com.google.common.collect.ImmutableMap;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.opensearch.action.support.master.AcknowledgedResponse;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.knn.KNNSingleNodeTestCase;
+import org.opensearch.knn.index.util.KNNEngine;
+import org.opensearch.knn.indices.Model;
+import org.opensearch.knn.indices.ModelDao;
+import org.opensearch.knn.indices.ModelMetadata;
+import org.opensearch.knn.indices.ModelState;
+
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.opensearch.knn.common.KNNConstants.INDEX_DESCRIPTION_PARAMETER;
+import static org.opensearch.knn.common.KNNConstants.SPACE_TYPE;
+
+public class KNNCreateIndexFromModelTests extends KNNSingleNodeTestCase {
+
+    public void testCreateIndexFromModel() throws IOException, InterruptedException {
+        // This test confirms that we can create an index from cluster metadata
+
+        // Create a model offline
+        String modelId = "test-model";
+        KNNEngine knnEngine = KNNEngine.FAISS;
+        SpaceType spaceType = SpaceType.L2;
+        int dimension = 3;
+
+        // "Train" a faiss flat index - this really just creates an empty index that does brute force k-NN
+        long vectorsPointer = JNIService.transferVectors(0, new float[0][0]);
+        byte [] modelBlob = JNIService.trainIndex(ImmutableMap.of(
+                INDEX_DESCRIPTION_PARAMETER, "Flat",
+                SPACE_TYPE, spaceType.getValue()), dimension, vectorsPointer,
+                KNNEngine.FAISS.getName());
+
+        // Setup model
+        ModelMetadata modelMetadata = new ModelMetadata(knnEngine, spaceType, dimension, ModelState.CREATED,
+                TimeValue.timeValueDays(3), "", "");
+
+        Model model = new Model(modelMetadata, modelBlob);
+
+        // Check that an index can be created from the model.
+        ModelDao modelDao = ModelDao.OpenSearchKNNModelDao.getInstance();
+
+        final CountDownLatch inProgressLatch = new CountDownLatch(1);
+
+        ActionListener<AcknowledgedResponse> mappingListener = ActionListener.wrap(response -> {
+            assertTrue(response.isAcknowledged());
+            inProgressLatch.countDown();
+        }, e -> fail("Unable to put mapping: " + e.getMessage()));
+
+        String indexName = "test-index";
+        String fieldName = "test-field";
+
+        modelDao.put(modelId, model, ActionListener.wrap(indexResponse -> {
+            createKNNIndex("test-index");
+            PutMappingRequest request = new PutMappingRequest(indexName).type("_doc");
+            request.source(fieldName, "type=knn_vector,model_id=" + modelId);
+            client().admin().indices().putMapping(request, mappingListener);
+        }, e ->fail("Unable to put model: " + e.getMessage())));
+
+        assertTrue(inProgressLatch.await(20, TimeUnit.SECONDS));
+    }
+}

--- a/src/test/java/org/opensearch/knn/index/KNNCreateIndexFromModelTests.java
+++ b/src/test/java/org/opensearch/knn/index/KNNCreateIndexFromModelTests.java
@@ -13,9 +13,8 @@ package org.opensearch.knn.index;
 
 import com.google.common.collect.ImmutableMap;
 import org.opensearch.action.ActionListener;
-import org.opensearch.action.admin.indices.mapping.put.PutMappingRequest;
-import org.opensearch.action.support.master.AcknowledgedResponse;
-import org.opensearch.common.unit.TimeValue;
+import org.opensearch.action.admin.indices.create.CreateIndexRequestBuilder;
+import org.opensearch.common.settings.Settings;
 import org.opensearch.knn.KNNSingleNodeTestCase;
 import org.opensearch.knn.index.util.KNNEngine;
 import org.opensearch.knn.indices.Model;
@@ -24,6 +23,8 @@ import org.opensearch.knn.indices.ModelMetadata;
 import org.opensearch.knn.indices.ModelState;
 
 import java.io.IOException;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
 
@@ -50,7 +51,7 @@ public class KNNCreateIndexFromModelTests extends KNNSingleNodeTestCase {
 
         // Setup model
         ModelMetadata modelMetadata = new ModelMetadata(knnEngine, spaceType, dimension, ModelState.CREATED,
-                TimeValue.timeValueDays(3), "", "");
+                ZonedDateTime.now(ZoneOffset.UTC).toString(), "", "");
 
         Model model = new Model(modelMetadata, modelBlob);
 
@@ -59,19 +60,36 @@ public class KNNCreateIndexFromModelTests extends KNNSingleNodeTestCase {
 
         final CountDownLatch inProgressLatch = new CountDownLatch(1);
 
-        ActionListener<AcknowledgedResponse> mappingListener = ActionListener.wrap(response -> {
-            assertTrue(response.isAcknowledged());
-            inProgressLatch.countDown();
-        }, e -> fail("Unable to put mapping: " + e.getMessage()));
-
         String indexName = "test-index";
         String fieldName = "test-field";
 
         modelDao.put(modelId, model, ActionListener.wrap(indexResponse -> {
-            createKNNIndex("test-index");
-            PutMappingRequest request = new PutMappingRequest(indexName).type("_doc");
-            request.source(fieldName, "type=knn_vector,model_id=" + modelId);
-            client().admin().indices().putMapping(request, mappingListener);
+            CreateIndexRequestBuilder createIndexRequestBuilder = client().admin().indices().prepareCreate(indexName)
+                    .setSettings(Settings.builder()
+                            .put("number_of_shards", 1)
+                            .put("number_of_replicas", 0)
+                            .put("index.knn", true)
+                            .build()
+                    ).addMapping(
+                            "_doc", ImmutableMap.of(
+                                    "properties", ImmutableMap.of(
+                                            fieldName, ImmutableMap.of(
+                                                    "type", "knn_vector",
+                                                    "model_id", modelId
+                                            )
+                                    )
+                            )
+                    );
+
+            client().admin().indices().create(createIndexRequestBuilder.request(),
+                    ActionListener.wrap(
+                            createIndexResponse -> {
+                                assertTrue(createIndexResponse.isAcknowledged());
+                                inProgressLatch.countDown();
+                            }, e -> fail("Unable to create index: " + e.getMessage())
+                    )
+            );
+
         }, e ->fail("Unable to put model: " + e.getMessage())));
 
         assertTrue(inProgressLatch.await(20, TimeUnit.SECONDS));

--- a/src/test/java/org/opensearch/knn/index/KNNQueryBuilderTests.java
+++ b/src/test/java/org/opensearch/knn/index/KNNQueryBuilderTests.java
@@ -32,6 +32,8 @@ import org.opensearch.common.xcontent.XContentParser;
 import org.opensearch.index.Index;
 import org.opensearch.index.mapper.NumberFieldMapper;
 import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.knn.indices.ModelDao;
+import org.opensearch.knn.indices.ModelMetadata;
 
 import java.io.IOException;
 
@@ -110,6 +112,32 @@ public class KNNQueryBuilderTests extends KNNTestCase {
         assertEquals(knnQueryBuilder.vector(), query.getQueryVector());
     }
 
+    public void testDoToQuery_FromModel() throws Exception {
+        float[] queryVector = {1.0f, 2.0f, 3.0f, 4.0f};
+        KNNQueryBuilder knnQueryBuilder = new KNNQueryBuilder("myvector", queryVector, 1);
+        Index dummyIndex = new Index("dummy", "dummy");
+        QueryShardContext mockQueryShardContext = mock(QueryShardContext.class);
+        KNNVectorFieldMapper.KNNVectorFieldType mockKNNVectorField = mock(KNNVectorFieldMapper.KNNVectorFieldType.class);
+        when(mockQueryShardContext.index()).thenReturn(dummyIndex);
+
+        // Dimension is -1. In this case, model metadata will need to provide dimension
+        when(mockKNNVectorField.getDimension()).thenReturn(-1);
+        String modelId = "test-model-id";
+        when(mockKNNVectorField.getModelId()).thenReturn(modelId);
+
+        // Mock the modelDao to return mocked modelMetadata
+        ModelMetadata modelMetadata = mock(ModelMetadata.class);
+        when(modelMetadata.getDimension()).thenReturn(4);
+        ModelDao modelDao = mock(ModelDao.class);
+        when(modelDao.getMetadata(modelId)).thenReturn(modelMetadata);
+        KNNQueryBuilder.initialize(modelDao);
+
+        when(mockQueryShardContext.fieldMapper(anyString())).thenReturn(mockKNNVectorField);
+        KNNQuery query = (KNNQuery)knnQueryBuilder.doToQuery(mockQueryShardContext);
+        assertEquals(knnQueryBuilder.getK(), query.getK());
+        assertEquals(knnQueryBuilder.fieldName(), query.getField());
+        assertEquals(knnQueryBuilder.vector(), query.getQueryVector());
+    }
 
     public void testDoToQuery_InvalidDimensions() {
         float[] queryVector = {1.0f, 2.0f, 3.0f, 4.0f};

--- a/src/test/java/org/opensearch/knn/index/codec/KNNCodecTestCase.java
+++ b/src/test/java/org/opensearch/knn/index/codec/KNNCodecTestCase.java
@@ -35,6 +35,7 @@ import org.opensearch.knn.index.JNIService;
 import org.opensearch.knn.index.KNNQuery;
 import org.opensearch.knn.index.KNNSettings;
 import org.opensearch.knn.index.KNNVectorFieldMapper;
+import org.opensearch.knn.index.KNNWeight;
 import org.opensearch.knn.index.SpaceType;
 import org.opensearch.knn.index.VectorField;
 import org.opensearch.knn.index.codec.KNN87Codec.KNN87Codec;
@@ -230,6 +231,7 @@ public class  KNNCodecTestCase extends KNNTestCase {
 
         Model mockModel = new Model(modelMetadata1, modelBlob);
         when(modelDao.get(modelId)).thenReturn(mockModel);
+        when(modelDao.getMetadata(modelId)).thenReturn(modelMetadata1);
 
         Settings settings = settings(CURRENT).put(MODEL_CACHE_SIZE_IN_BYTES_SETTING.getKey(), 10).build();
         ClusterSettings clusterSettings = new ClusterSettings(settings,
@@ -251,9 +253,6 @@ public class  KNNCodecTestCase extends KNNTestCase {
 
         FieldType fieldType = new FieldType(KNNVectorFieldMapper.Defaults.FIELD_TYPE);
         fieldType.putAttribute(KNNConstants.MODEL_ID, modelId);
-        fieldType.putAttribute(KNNConstants.KNN_ENGINE, KNNEngine.FAISS.getName());
-        fieldType.putAttribute(KNNConstants.SPACE_TYPE, spaceType.getValue());
-        fieldType.putAttribute(KNNConstants.DIMENSION, String.valueOf(dimension));
         fieldType.freeze();
 
         // Add the documents to the index
@@ -277,6 +276,7 @@ public class  KNNCodecTestCase extends KNNTestCase {
         writer.close();
 
         // Make sure that search returns the correct results
+        KNNWeight.initialize(modelDao);
         NativeMemoryLoadStrategy.IndexLoadStrategy.initialize(createDisabledResourceWatcherService());
         float [] query = {10.0f, 10.0f, 10.0f};
         IndexSearcher searcher = new IndexSearcher(reader);

